### ISSUE TITLE
prefer_interpolation_to_compose_strings is applied too broadly

### ISSUE THREAD

**Issue by aemino** [dart-lang/linter#792]
Hi, it's me again. :)
I'm doing something like this frequently in my code:
```dart
Route get route => client.api + 'guilds/$id';
```
This triggers the `prefer_interpolation_to_compose_strings` rule.  However, `client.api` is a `Route` instance, and not a string instance.  The rule interprets this as a string being composed; thus it lints this case.

**Comment by alexeieleusis** [comment]
Is this the same operator that returned `this`?

**Comment by aemino** [comment]
Indeed, it is.

**Comment by alexeieleusis** [comment]
No. I am out of context to be honest.

At first sight looks like at runtime `toString()` will be invoked on `client.api` unless it has an overload on the `+` operator. Is that the case @aemino ? If that's the case, making the lint understand that might make it slow.

In the case `toString()` is invoked, the lint is working as expected, I don't know if interpolating would increase readability, but programmatically distinguishing those cases can be tricky.

**Comment by aemino** [comment]
Indeed, @alexeieleusis.  `client.api` is an instance of a class called `Route`, which has its `+` operator overridden.  If I recall correctly, this issue brings to light the fact that this lint triggers when either side of a `+` operation is a string.  If I am not mistaken, it makes more sense for this lint to only trigger when both the left and right elements are strings.

**Comment by aemino** [comment]
What else could the lint apply to?  If I'm not mistaken, Dart 2 prevents users from using `String`'s `+` operator on any other types besides `String`.

**Comment by alexeieleusis** [comment]
I thought it implicitly invoked toString on the right hand operand.

**Comment by aemino** [comment]
Using Dart `2.0.0-dev.65.0`, this code...
```dart
print('hello' + 5);
```
yields this error:
`The argument type 'int' can't be assigned to the parameter type 'String'.`

**Comment by zoechi** [comment]
@alexeieleusis only with interpolation `print('hello${5}');`

**Comment by aemino** [comment]
While we're on the subject of this rule being applied too broadly, I would also like to suggest that there be a complexity and/or length criterion.  I notice that occasionally in my code, I need to concatenate a string with another string that is created from a complex functional statement.  It would look rather ridiculous if I stuck all of that code inside an interpolation.

**Comment by zoechi** [comment]
You can still assign the result of the code to a variable and use the variable in interpolation.

**Comment by bwilkerson** [comment]
You could also create a `StringBuilder` and compose the string that way. I believe that it's slightly more efficient than composing multiple strings (though I don't know whether you're talking about performance critical code).

**Comment by aemino** [comment]
I'm aware that these are options, but they don't really work when you're dealing with "functional" code.  For example, when writing builder generator classes, it seems that writing code with copious numbers of lambdas is the norm.
